### ISSUE TITLE
Support Creation Semantics Needed by Bucket Types

### ISSUE THREAD

**Issue by jrwest** [basho/riak_core#376]
The bucket types implementation added in #357 does not properly enforce the expected creation semantics  -- that it cannot result in siblings. This is because #363 only includes eventually consistent writes. Creation of bucket types must be conflict-free to ensure there are no ambiguities when features like strong-consistency and CRDTs use them to specify the types of data stored under buckets in a type.  

Some (hand-wavey) proposals that have been thrown out:
1. serialize through claimaint
2. some sort of "seen" set overlaid on cluster metadata
3. use strongly consistent store that will be added to riak core as part of strong consistency in riak work

1 & 3 seem to be the most likely and easiest to implement candidates.

In addition to the work in cluster metadata, bucket types should be converted to using the changes before closing this issue.

**Comment by jrwest** [comment]
a first pass has been implemented here https://github.com/basho/riak_core/commit/b4d9525b97c499f8b2397cdfeb3550318d82dac1

**Comment by jrwest** [comment]
a second pass has been implemented here: https://github.com/basho/riak_core/tree/jrw-bucket-type-invariant

**Comment by jrwest** [comment]
also related https://github.com/basho/riak_kv/pull/673 and https://github.com/basho/riak_test/pull/394

**Comment by jrwest** [comment]
Also related and can be reviewed together w/ the above PRs (although they don't necessarily need to be): https://github.com/basho/riak_kv/pull/677 https://github.com/basho/riak/pull/395

**Comment by jrwest** [comment]
The web of PRs above can be a bit hard to follow, so here is some more detail:

#402 is the basis for these changes and must be reviewed along-side https://github.com/basho/riak_kv/pull/673. The bucket types test has been updated as well in https://github.com/basho/riak_test/pull/394. This should help w/ review of the previous two PRs. In addition riak-admin support has been added. This is done in https://github.com/basho/riak_kv/pull/677 and https://github.com/basho/riak/pull/395 which also need to be reviewed together but may be separate from the previous group of PRs.

**Comment by Vagabond** [comment]
Can we close this now?

**Comment by jrwest** [comment]
ah yes, sorry, got lost in the shuffle.
